src/sequential.cpp

```diff
@@ -137,10 +138,19 @@ void Sequential::init_output_state_buffer()
  */
 {
     if (this->device.compare("cpu") == 0) {
-        this->output_z_buffer = std::make_shared<BaseHiddenStates>(
-            this->z_buffer_size, this->z_buffer_block_size);
-        this->input_z_buffer = std::make_shared<BaseHiddenStates>(
-            this->z_buffer_size, this->z_buffer_block_size);
+        if (this->layers[0]->get_layer_type() == LayerType::SLSTM) {
```

Comment:
I tried to add this, but it does not work because when using cuda, the LSTMCuda layer is used, then its LayerType is still LSTM so that I cannot make the condition:
`(this->layers[0]->get_layer_type() == LayerType::SLSTM)`

Reply:
here is the full code
```c++
void Sequential::init_output_state_buffer()
/*
 */
{
    if (this->device.compare("cpu") == 0) {
        if (this->layers[0]->get_layer_type() == LayerType::SLSTM) {
            this->output_z_buffer = std::make_shared<SmoothingHiddenStates>(
                this->z_buffer_size, this->z_buffer_block_size,
                this->num_samples);
            this->input_z_buffer = std::make_shared<SmoothingHiddenStates>(
                this->z_buffer_size, this->z_buffer_block_size,
                this->num_samples);
        } else {
            this->output_z_buffer = std::make_shared<BaseHiddenStates>(
                this->z_buffer_size, this->z_buffer_block_size);
            this->input_z_buffer = std::make_shared<BaseHiddenStates>(
                this->z_buffer_size, this->z_buffer_block_size);
        }
        this->temp_states = std::make_shared<BaseTempStates>(
            this->z_buffer_size, this->z_buffer_block_size);
    }
#ifdef USE_CUDA
    else if (this->device.compare("cuda") == 0) {
        if (this->layers[0]->get_layer_type() != LayerType::SLSTM) {
            this->output_z_buffer = std::make_shared<HiddenStateCuda>(
                this->z_buffer_size, this->z_buffer_block_size);
            this->input_z_buffer = std::make_shared<HiddenStateCuda>(
                this->z_buffer_size, this->z_buffer_block_size);
            this->temp_states = std::make_shared<TempStateCuda>(
                this->z_buffer_size, this->z_buffer_block_size);
        } else {
            throw std::invalid_argument(
                "Error in file: " + std::string(__FILE__) +
                " at line: " + std::to_string(__LINE__) +
                ". Smoothing feature does not support CUDA");
        }
    }
#endif
    else {
        throw std::invalid_argument("Error in file: " + std::string(__FILE__) +
                                    " at line: " + std::to_string(__LINE__) +
                                    ". Invalid device: [" + this->device + "]");
    }
}
```

Reply:
I tried. It gives errors in `if (this->layers[0]->get_layer_type() != LayerType::SLSTM)` because `this->layers[0]->get_layer_type()` returns `LSTM` when using cuda, and `this->layers[0]->get_layer_name()` returns `LSTMcuda`

Reply:
This is a good idea, but I tried, it still gives errors. Could not get in `else {
            throw std::invalid_argument(
                "Error in file: " + std::string(__FILE__) +
                " at line: " + std::to_string(__LINE__) +
                ". Smoothing feature does not support CUDA");
        }`

Reply:
Note that I added your suggested full code above for `void Sequential::init_output_state_buffer()`, but it still does not work throw the message when using cuda. Instead, it still gives errors.

Reply:
oh I see. It is because you missed this [line](https://github.com/lhnguyen102/cuTAGI/blob/main/include/linear_layer.h#L144). The code will be segment defaut before reaching that point. I will take care of it with another PR